### PR TITLE
Use AccountsService User Icon

### PR DESCRIPTION
**Pull request by Artoria2e5**
This set of commits is a part of #439. I cleaned them up by squashing fixups and rewording.

Hope this will be good enough. // CC @xiangzhai

### CODE REVIEW

**Comment by Arthur2e5** [comment]
By the way I am not really sure about the headers used, so if someone has some environment to test…

**Comment by xiangzhai** [comment]
:+1:

**Comment by Arthur2e5** [comment]
CONFLICT?!

**Comment by xiangzhai** [comment]
you are git Jedi Master ;-)

**Comment by jleclanche** [comment]
The commits can be squashed together. Do this with `git rebase -i <upstream>/master` and replacing the last two lines' `pick` with `fixup`.

**Comment by Arthur2e5** [comment]
I don't consider squashing / fixup-ing the first two a good idea.

**Comment by jleclanche** [comment]
Why? all your commits do is fix the first commit.

**Comment by Arthur2e5** [comment]
I thought I should still keep traces of some old API versions.

**Comment by jleclanche** [comment]
Please rebase against master.

**Comment by tpgxyz** [comment]
Hi,

i've prepared a patch for it
https://abf.io/openmandriva/sddm/blob/master/sddm-0.14.0-add-support-to-QtAccountsService.patch


Unfortunately it does not compile (with updated to latest git qtaccountservice - btw would be nice to do a release :+1: )
../src/greeter/UserModel.cpp:88:45: warning: unused variable 'asUserAccount' [-Wunused-variable]
            QtAccountsService::UserAccount *asUserAccount = asAccountsManager->findUserByName(user->name);
                                            ^
../src/greeter/UserModel.cpp:124:12: error: use of undeclared identifier 'asUserAccount'
                else if (asUserAccount && QFile::exists(asUserAccount->iconFileName()))
                         ^
../src/greeter/UserModel.cpp:124:43: error: use of undeclared identifier 'asUserAccount'
                else if (asUserAccount && QFile::exists(asUserAccount->iconFileName()))
                                                        ^
../src/greeter/UserModel.cpp:125:20: error: use of undeclared identifier 'asUserAccount'
                    user->icon = asUserAccount->iconFileName(); // accountservice user face

**Comment by Arthur2e5** [comment]
Hey, I have squashed & rebased the changes as well as addressed your suggestions. Please merge this PR if it looks OK now.

File: src/greeter/UserModel.cpp [hunk 1]
```diff
@@ -79,12 +84,27 @@ namespace SDDM {
             user->homeDir = QString::fromLocal8Bit(current_pw->pw_dir);
             user->uid = int(current_pw->pw_uid);
             user->gid = int(current_pw->pw_gid);
+#if HAVE_QTACCOUNTSSERVICE
+            AccountsService::UserAccount *asUserAccount = asAccountsManager->findUserByName(user->name);
```
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: Ugh, where is my UID edit? Wait....

**Comment by tpgxyz** [comment]
@Arthur2e5 @xiangzhai 

Looks like your code does not compile:

/UserModel.cpp.o -c ../src/greeter/UserModel.cpp
../src/greeter/UserModel.cpp:60:33: error: use of undeclared identifier 'AccountsService'; did you mean 'QtAccountsService'?
        asAccountsManager = new AccountsService::AccountsManager;
                                ^~~~~~~~~~~~~~~
                                QtAccountsService
/usr/include/QtAccountsService/qtaccountsservice/accountsmanager.h:37:11: note: 'QtAccountsService' declared here
namespace QtAccountsService {
          ^
../src/greeter/UserModel.cpp:88:13: error: use of undeclared identifier 'AccountsService'; did you mean 'QtAccountsService'?
            AccountsService::UserAccount *asUserAccount = asAccountsManager->findUserById(user->uid);
            ^~~~~~~~~~~~~~~
            QtAccountsService
/usr/include/QtAccountsService/qtaccountsservice/accountsmanager.h:37:11: note: 'QtAccountsService' declared here
namespace QtAccountsService {
          ^
2 errors generated.

Full log output can be found here http://file-store.openmandriva.org/api/v1/file_stores/6f0e359e361491b65936daf6f62b81973b41e4d1.log?show=true

**Comment by Arthur2e5** [comment]
Sounds like a correct guess by the compiler. Rebase and force-push in progress..

**Comment by tpgxyz** [comment]
@Arthur2e5 Yes LLVM/clang is smart :)

**Comment by tpgxyz** [comment]
@Arthur2e5  @xiangzhai  Still does not compile.

../src/greeter/UserModel.cpp:60:27: error: assigning to 'QtAccountsService::AccountsManager *' from incompatible type 'QtAccountsService::AccountsManager **'; dereference with *
        asAccountsManager = new QtAccountsService::AccountsManager
                          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

File: src/greeter/UserModel.cpp [hunk 2]
```diff
@@ -54,6 +56,9 @@ namespace SDDM {
         const QString facesDir = mainConfig.Theme.FacesDir.get();
         const QString defaultFace = QStringLiteral("%1/.face.icon").arg(facesDir);
 
+#if HAVE_QTACCOUNTSSERVICE
+        asAccountsManager = new QtAccountsService::AccountsManager;
```
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: I guess I should do `asAccountsManager = &(new..);` here. @xiangzhai

File: src/greeter/UserModel.cpp [hunk 3]
```diff
@@ -120,6 +140,13 @@ namespace SDDM {
 
     UserModel::~UserModel() {
         delete d;
+
+#if HAVE_QTACCOUNTSSERVICE
+        if (asAccountsManager) {
```
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: Smells like a useless null check.

File: src/greeter/UserModel.cpp [hunk 4]
```diff
@@ -120,6 +140,13 @@ namespace SDDM {
...
+            delete asAccountsManager;
+            asAccountsManager = nullptr;
```
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: Smells like a useless assignment since the instance is no longer there.

File: src/greeter/UserModel.cpp (hunk 2 above)
Comment (xiangzhai) [inline src/greeter/UserModel.cpp]: https://en.wikipedia.org/wiki/New_and_delete_(C%2B%2B)

File: src/greeter/UserModel.cpp (hunk 3 above)
Comment (xiangzhai) [inline src/greeter/UserModel.cpp]: asAccountsManager reference might be failed to new ;-)

File: src/greeter/UserModel.cpp (hunk 4 above)
Comment (xiangzhai) [inline src/greeter/UserModel.cpp]: it is a just good programming habit ;-)

File: src/greeter/UserModel.cpp (hunk 3 above)
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: deleting a `nullptr` should be always fine...

File: src/greeter/UserModel.cpp (hunk 2 above)
Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: Hmmmm. `/me is confused`.

Comment (Arthur2e5) [inline src/greeter/UserModel.cpp]: Hmmmm. Perhaps I should try to un-markdown @tpgxyz's output first.

```
../src/greeter/UserModel.cpp:60:27: error: assigning to 'QtAccountsService::AccountsManager *' from incompatible type 'QtAccountsService::AccountsManager **'; dereference with *
```

(Yeah, try to wrap your paste with ```` ``` ```` next time.)

**Comment by tpgxyz** [comment]
Hi, any news on this @Arthur2e5  @xiangzhai :)

**Comment by xiangzhai** [comment]
Hi @tpgxyz what about Plasma 5.7`s all-new login screen design https://www.kde.org/announcements/plasma-5.6.95.php KDM will come back? or just SDDM will create such Theme? it seems using user's Accounts icon ;-)

**Comment by tpgxyz** [comment]
@xiangzhai I do not see anything related to sddm new theme or any new login manager in plasma 5.6.95 sources

**Comment by greyhare** [comment]
@xiangzhai That just looks like an update of the KDE Breeze theme for SDDM, not any change in login manager.

**Comment by plfiorini** [comment]
it's still SDDM, they just updated the graphics.
how can you tell it's using users accounts? KDE is probably using the ancient KUser api.

**Comment by xiangzhai** [comment]
YES! @plfiorini we tried to, but ... https://mail.kde.org/pipermail/plasma-devel/2015-July/042766.html

**Comment by xiangzhai** [comment]
> they just updated the graphics.

@plfiorini ArchLinux users concerned about the new graphics...  https://bbs.archlinux.org/viewtopic.php?pid=1635352#p1635352

so I just backport the components/artwork/background.png https://twitter.com/xiangzhai/status/745800173348020225

File: src/greeter/UserModel.cpp (hunk 2 above)
Comment (Artoria2e5) [inline src/greeter/UserModel.cpp]: So why does `new QtAccountsService::AccountsManager` give `QtAccountsService::AccountsManager **`…

Comment (xiangzhai) [inline src/greeter/UserModel.cpp]: you can use gdb to print it out ;-)

Comment (Artoria2e5) [inline src/greeter/UserModel.cpp]: Using gdb on something that doesn't even compile, hmmmmmm forget it.

**Comment by davidedmundson** [comment]
> how can you tell it's using users accounts? KDE is probably using the ancient KUser api.
KDE now uses AccountsService in KUser as of 2 weeks ago.

Rather than going via the daemon we just hardcode /var/lib/AccountsService/icons/$name
The daemon is useful for setting, but for reading it's not adding any value. It's sort of a lazy solution, but equally quite sensible. 

(Also I've only just seen that I was mentioned in https://github.com/hawaii-desktop/qtaccountsservice/issues/7 but I didn't see it because I'm not subscribed to hawaii-desktop things sorry about that)

**Comment by Pointedstick** [comment]
This can be closed now; support went in today with https://github.com/sddm/sddm/commit/6015d10fe771614d38026d37ceb26025dd711a28.

**Pull request closed without merging**